File: step3_consolidate_data.py
from __future__ import division

import os

import numpy as np
import pandas as pd


def _prep_conn_df(df):
    """
    >>> import pandas as pd
    >>> import numpy as np
    >>> data = np.random.rand(3, 3)
    >>> from step3_consolidate_data import _prep_conn_df
    >>> ret = _prep_conn_df(pd.DataFrame(data))
    >>> expected_rows = ['0_1', '0_2', '1_2']
    >>> assert np.all(np.sort(ret.index) == expected_rows)
    >>> get_val = lambda c1, c2: ret[ret.index=='%s_%s' % (c1, c2)]['value'].iloc[0]
    >>> assert get_val(0, 1) == data[0][1]
    >>> assert get_val(0, 2) == data[0][2]
    >>> assert get_val(1, 2) == data[1][2]

    """

    num_rois = len(df.columns)

    tmp = df.copy()
    tmp['roi1'] = tmp.index.copy()

    ret = pd.melt(tmp, id_vars=['roi1'], var_name='roi2')
    ret = ret[ret['roi1'] < ret['roi2']]

    ret.index = ret.apply(lambda r: '%s_%s' % (r['roi1'], r['roi2']), axis=1)
    del ret['roi1']
    del ret['roi2']

    num_rows = (num_rois * (num_rois-1))/2
    assert ret.shape == (num_rows, 1)

    return ret


def load_conn_csv(csv_path):
    df = _prep_conn_df(pd.read_csv(csv_path, header=None))
    df = df.T
    df.index = [os.path.basename(csv_path)[:-4]]
    return df


def load_lesion_csv(csv_path):
    return pd.read_csv(csv_path, header=None).T


def consolidate_x(csv_dir, dest_path, load_fn):
    fs = os.listdir(csv_dir)
    num_fs = len(fs)

    def load(csv, wait=dict(wait=0)):
        wait['wait'] += 1
        print "%s out of %s" % (wait['wait'], num_fs)
        return load_fn(os.path.join(csv_dir, csv))

    dfs = pd.concat([load(c) for c in fs], axis=0)
    assert dfs.shape[0] == num_fs

    if dest_path:
        dfs.to_csv(dest_path)

    return dfs


def consolidate_conn_x(csv_dir="data/step1/conn", dest_path="data/step3/conn/X.csv"):
    print("conn")
    consolidate_x(csv_dir, dest_path, load_conn_csv)


def consolidate_lesion_x(csv_dir="data/step1/lesion", dest_path="data/step3/lesion/X.csv"):
    print("lesion")
    consolidate_x(csv_dir, dest_path, load_lesion_csv)


def consolidate_x_with_y(y_dir="data/step2/",
                x_src="data/step3/X.csv",
                dest_dir="data/step3/"):
    x = pd.read_csv(x_src, index_col=0)

    ret = dict()
    for yo in ['adw', 'atw']:
        y_src = os.path.join(y_dir, yo+'_outcomes.csv')
        y = pd.read_csv(y_src)
        full = pd.merge(x, y, right_on='id', left_index=True, how='inner')

        max_num_rows = np.min((x.shape[0], y.shape[0]))
        num_cols = x.shape[1] + y.shape[1]

        assert full.shape[0] <= max_num_rows
        assert full.shape[1] == num_cols

        if dest_dir:
            dest_path = os.path.join(dest_dir, 'full_%s.csv' % yo)
            full.to_csv(dest_path, index=False)

        ret[yo] = full

    return ret


def consolidate_lesion_x_with_y(x_src="data/step3/lesion/X.csv",
                       dest_dir="data/step3/lesion/"):
    consolidate_x_with_y(x_src=x_src, dest_dir=dest_dir)


def consolidate_conn_x_with_y(x_src="data/step3/conn/X.csv",
                       dest_dir="data/step3/conn/"):
    consolidate_x_with_y(x_src=x_src, dest_dir=dest_dir)


def run_all():
    consolidate_conn_x()
    consolidate_conn_x_with_y()

    consolidate_lesion_x()
    consolidate_lesion_x_with_y()


def _test():
    import doctest
    doctest.testfile("step3_consolidate_data.py")

File: step3ConsolidateData.jl
using DataFrames
using Lazy
using Memoize


loadRoiNames() = Vector{Symbol}(map(symbol, readtable("data/jhu_coords.csv")[:name]))


@memoize getRoiIx(roi::Symbol) = findfirst(e -> e == roi, loadRoiNames())


getId(csv_path::AbstractString) = basename(csv_path)[1:end-4]


function loadConnCsv(csv_path, rois=loadRoiNames())
  roi_x_roi = readtable(csv_path, header=false, names=rois)

  num_rois = length(rois)

  @assert size(roi_x_roi) == (num_rois, num_rois)

  roi_x_roi[:left] = rois
  roi_long = rename!(melt(roi_x_roi, :left), :variable, :right)
  roi_long[:left_ix] = map(getRoiIx, roi_long[:left])
  roi_long[:right_ix] = map(getRoiIx, roi_long[:right])

  roi_long = roi_long[roi_long[:left_ix] .> roi_long[:right_ix], [:left, :right, :value]]
  conn_names::Vector{Symbol} = map(eachrow(roi_long)) do r
    symbol(r[:left], "_to_" , r[:right])
  end
  conn_values = Any[[v] for v in roi_long[:value]]

  ret = DataFrame(conn_values, conn_names)
  ret[:id] = getId(csv_path)
  ret
end


function loadLesionCsv(csv_path, rois=loadRoiNames())
  data::Vector{Any} = Any[[i] for i in readcsv(csv_path)]
  @assert length(rois) == length(data)

  ret = DataFrame(data, rois)
  ret[:id] = getId(csv_path)
  ret
end


typealias OptDir Nullable{ASCIIString}


function consolidateX(load_fn::Function,
                       csv_dir::AbstractString;
                       dest_dir::OptDir = OptDir())
  fs = readdir(csv_dir)
  num_fs = length(fs)

  curr_load::Int64 = 0
  function load(csv::AbstractString)
    curr_load += 1
    println("$(curr_load) out of $(num_fs)")
    load_fn(joinpath(csv_dir, csv))
  end

  consolidated_df::DataFrame = vcat(DataFrame[load(c) for c in fs])
  @assert size(consolidated_df, 1) == num_fs

  if !isnull(dest_dir)
    dest_path = joinpath(get(dest_dir), "X.csv")
    writetable(dest_path, consolidated_df)
  end

  consolidated_df
end


function consolidateConnX(csv_dir="data/step1/conn";
                            dest_dir::OptDir=OptDir())
  println("conn")
  rois = loadRoiNames()

  consolidateX(csv_dir, dest_dir=dest_dir) do csv_path
    loadConnCsv(csv_path, rois)
  end
end


function consolidateLesionX(csv_dir="data/step1/lesion";
                              dest_dir::OptDir=OptDir())
  println("lesion")
  rois::Vector{Symbol} = loadRoiNames()

  consolidateX(loadLesionCsv, csv_dir, dest_dir=dest_dir)
end


function calcWpsDiff!(y::DataFrame, outcome::ASCIIString)
  ending = symbol('_', outcome[end-1:end])
  addEnding(s) = symbol(s, ending)

  se_second_look_up = Dict(addEnding(:se_elvis) => 38,
    addEnding(:se_mlk) => 44,
    addEnding(:se_sm) => 43)

  num_rows = size(y, 1)

  naZeros(k::Symbol, v) = [isna(i) ? 0. : v for i in y[k]]
  naZeros(k::Symbol) = [isna(i) ? 0. : i for i in y[k]]

  se_total_seconds = reduce(zeros(num_rows), keys(se_second_look_up)) do acc, k
    acc .+ naZeros(k, se_second_look_up[k])
  end

  se_o_wps = symbol(:se_, outcome, :_wps)
  pd_o_wps = symbol(:pd_, outcome, :_wps)
  pd_o_wps_picnic = symbol(:pd_, outcome, :_wps, :_picnic)

  y[se_o_wps] = sum(map(naZeros, se_second_look_up |> keys))./se_total_seconds

  y[pd_o_wps] = y[symbol(:pd_, outcome)]./120.
  y[pd_o_wps_picnic] = y[symbol(:pd_, outcome)]./y[:picnic_seconds]

  o_diff_wps = symbol(outcome, :_diff_wps)
  o_diff_wps_picnic = symbol(o_diff_wps, :_picnic)

  y[o_diff_wps] = y[se_o_wps] - y[pd_o_wps]
  y[o_diff_wps_picnic] = y[se_o_wps] - y[pd_o_wps_picnic]

  y
end


function consolidateXwithY(consolidated_x::DataFrame,
                              y_dir="data/step2/";
                              dest_dir::OptDir=OptDir())

  @assert :id in names(consolidated_x)

  ret = Dict()
  meta = @> y_dir joinpath("meta.csv") readtable
  picnic_seconds = @> y_dir joinpath("picnic_seconds.csv") readtable

  idjoin(a, b) = join(a, b, on=:id)

  for yo in ["adw", "atw"]
    y_src = joinpath(y_dir, "$(yo)_outcomes.csv")
    y = readtable(y_src)

    full = @> consolidated_x begin
      idjoin(y)
      idjoin(meta)
      idjoin(picnic_seconds)
    end

    max_num_rows = @> size(consolidated_x, 1) min(size(y, 1))
    num_cols = size(consolidated_x, 2) + size(y, 2) + size(meta, 2) + size(picnic_seconds, 2) - 3

    @assert size(full, 1) <= max_num_rows
    @assert size(full, 2) == num_cols

    calcWpsDiff!(full, yo)

    @>> full begin
      names
      filter(c -> contains("$c", "diff"))
      map(c -> full[symbol(c, :_binary)] = full[c] .> 0.)
    end

    if !isnull(dest_dir)
      @> dest_dir get joinpath("full_$(yo).csv") writetable(full)
    end

    ret[yo] = full
  end

  ret
end


function runAll()
  lesion_dir = "data/step3/lesion"
  conn_dir = "data/step3/conn"

  X_lesion = consolidateLesionX(dest_dir=OptDir(lesion_dir))
  consolidateXwithY(X_lesion, dest_dir=OptDir(lesion_dir))

  X_conn = consolidateConnX(dest_dir=OptDir(conn_dir))
  consolidateXwithY(X_conn, dest_dir=OptDir(conn_dir))
end
